### ISSUE TITLE
space:format allows to set is_nullable = true on primary key field part

### ISSUE THREAD

**Issue by no1seman** [tarantool/tarantool#5607]
Tarantool version:
All versions (1.10.x, 2.x)
OS version:
No matter
Bug description:
space:format() allows to change field's 'is_nullable=true' property for field  that is a part of primary key (see the code below)
Steps to reproduce:
```lua
local yaml = require('yaml')

local space_name = 'megaspace'

local initial_format = {
    {name = 'customer_id', type = 'string', is_nullable = false},
    {name = 'id', type = 'string', is_nullable = false},
}

local new_format = {
    {name = 'customer_id', type = 'string', is_nullable = false},
    {name = 'id', type = 'string', is_nullable = true},
}

local function init_space(space)
    local _space = box.schema.space.create(space, {if_not_exists = true})
    _space:format(initial_format)
    _space:create_index('pk', {unique = true, parts = {'id'}, if_not_exists = true})
    return _space
end

local function change_format(space)
    box.space[space]:format(new_format)
end

local function add_data(data)
    box.space[space_name]:insert(data)
end

box.cfg({})

--create space
local space = init_space(space_name)
print(yaml.encode(space:format()))

-- add some data
add_data(box.tuple.new({'VIP1', '1'}))
print(yaml.encode(space:select()))

change_format(space_name)
print(yaml.encode(space:format()))
```

Optional (but very desirable):
* coredump
* backtrace
* netstat

**Comment by mary3000** [comment]
@no1seman thanks for the report!

Actually it is the desirable behaviour. See [warning](https://www.tarantool.io/en/doc/2.2/reference/reference_lua/box_space/#lua-function.space_object.create_index) in create_index section in docs:
> It is legal to create multiple indexes for the same field with different is_nullable values, or to call space_object:format() with a different is_nullable value from what is used for an index. When there is a contradiction, the rule is: null is illegal unless is_nullable=true for every index and for the space format.

I.e. in your example format allows nullable values, while primary key disallows (`space.index.pk.is_nullable` is null), and the final result is to forbid inserting nullable values. You can try it and get an error, as expected (`add_data({box.NULL, '1'})`)

**Comment by no1seman** [comment]
@mary3000 Seems that my example code is not relevant, here is a little bit modified:

```lua
local yaml = require('yaml')

local space_name = 'megaspace'

local initial_format = {
    {name = 'customer_id', type = 'string', is_nullable = false},
    {name = 'id', type = 'string', is_nullable = false},
}

local new_format = {
    {name = 'customer_id', type = 'string', is_nullable = false},
    {name = 'id', type = 'string', is_nullable = true},
}

local function init_space(space)
    local _space = box.schema.space.create(space, {if_not_exists = true})
    _space:format(initial_format)
    _space:create_index('pk', {unique = true, parts = {'id'}, if_not_exists = true})
    return _space
end

local function change_format(space)
    box.space[space]:format(new_format)
end

local function add_data(data)
    box.space[space_name]:insert(data)
end

box.cfg({})

if not box.space[space_name] then
    --create space
    print('Init space:')
    local space = init_space(space_name)
    print(yaml.encode(space:format()))

    -- add some data
    print('Add some data:')
    add_data(box.tuple.new({'VIP1', '1'}))
    print(yaml.encode(space:select()))

    print('Make "id" field nullable:')
    change_format(space_name)
    print(yaml.encode(space:format()))
else
    print('Add first tuple with "id" field==nil:')
    add_data(box.tuple.new({'VIP2', 'nil'}))
    print('Add second tuple with "id" field==nil:')
    add_data(box.tuple.new({'VIP3', 'nil'}))
end
```

1. Copy provided code to init.lua file.
2. Run: tarantool init.lua
3. Exit loop
4. Run once more: tarantool init.lua

After 4-th step you will get an error:
ER_TUPLE_FOUND: Duplicate key exists in unique index 'pk' in space 'megaspace'

Yes, you are right that it's a logical error and that behavior is fully according to documentation, but seems that space:format() must prevent setting PK field is_nullable, because in that case any input data checks based on space:format() will be positive but adding more that one tuple with nil will cause critical error. My point is that core logic must prevent such an logic errors and space:format() must perform check based on space index settings.

**Comment by Gerold103** [comment]
This contradicts with the whole idea why it was allowed in the first place. For the sake of incremental DDL. So as you could switch nullable on and off in separate DDL transactions without messing with transactional multistatement DDL which is still unstable and can crash sometimes.

**Comment by sharonovd** [comment]
- primary key cannot be altered
 - there's no sense in allowing nullable fields in pk
 
I think it's to make an exception here

**Comment by sharonovd** [comment]
Please feel free to involve me in triage process

**Comment by Gerold103** [comment]
>there's no sense in allowing nullable fields in pk

It is allowed, in case you intend to store only one NULL here.
Additionally, by forbidding it now you will break the backward compatibility in case somebody relied on this being allowed.

**Comment by Gerold103** [comment]
Please, feel free to add me to all the discussion so as I could block it before somebody would rush to make a patch.

**Comment by kostja** [comment]
It should not be allowed. The ansi way of a table with NULL key is to have no primary key. Primary key can't have nulls.

**Comment by sharonovd** [comment]
Moreover, the following snippet gives an error:
```
tarantool> t:format({{name='a', is_nullable=true},{name='b', is_nullable=true}})
---
...

tarantool> t:create_index('pk', {parts={'a', 'b'}})
---
- error: Primary index of space 'test' can not contain nullable parts

```

**Comment by Egor2001** [comment]
Seems that even our tests rely on the ability to set `is_nullable` to true for the parts of primary key.
For instance, `engine/ddl.test.lua` has such annotation:
```lua
i = s:create_index('primary', { parts = {'field1'} })
format[1].is_nullable = true
-- The format is allowed since in primary index parts
-- is_nullable is still set to false.
s:format(format)
```
Also the current behaviour for the sample code reproducing this bug is not `ER_TUPLE_FOUND` at the second insert but `ER_FIELD_MISSING` at the attempt to insert the first null value.
The possible patch for this issue and tests affected  are at 3de8eb8, but I'm sure that we need a really strong reason to change the current behaviour.

**Comment by alyapunov** [comment]
1. I could not reproduce `ER_TUPLE_FOUND` too (*). Perhaps we should bisect to find out when if was changed.
 So as I see if a primary index exists, it doesn't accept box.NULL or absent fields. And it seems to be correct.

2. Now `create_index` method takes field properties from space format, unless a property is set explicitly (**). No further changes in format will affect that index. That work not only for `is_nullable` but also for field names. In other words, now space format is independent from indexes. Theoretically we could make them dependent and rebuild indexes if format is changes, but that would be an epic issue.

3. Having indexes are independent of format, I'm sure that format should also be independent of index definitions. (And now it's already so). That would make lots of rules and limitations. If a user creates an index without specifying format - should the format be created? Same of formats exists but only contains only names (without types)? Should we restore previous format if the index is altered? That would be a huge RFC. 

4. Now a tuple in space must comply both format and index requirements. Perhaps we could make that the cumulative format getter in lua.

So my opinion - I don't see a critical mistake that tarantool makes. Both format and index defs work exactly as described in doc. If we want to make format and indexes dependent - we should not take only `is_nullable` into consideration, and create RFC about it.

<details>
<summary>(*) I have the following error</summary>

```
Add first tuple with "id" field==nil:
2021-06-28 14:32:52.970 [13876] main tuple_format.c:909 E> ER_FIELD_TYPE: Tuple field 2 (id) type does not match one required by operation: expected string, got nil
2021-06-28 14:32:52.970 [13876] main F> fatal error, exiting the event loop
```
</details>
<details>
<summary>(**) All possible variants</summary>

| pk definition \ space format | is_nullable is not specified | is_nullable is false | is_nullable is true |
|--|--|--|--|
| **is_nullable is not specified** | allowed | allowed | forbidden |
| **is_nullable is false** | allowed | allowed | allowed |
| **is_nullable is true** | forbidden | forbidden | forbidden |
</details>

**Comment by Gerold103** [comment]
I was asked to repeat a summary of my opinion on this ticket.

TLDR: not worth doing.

The main reason is that it is done intentionally. And it is even documented. You are allowed to have different `is_nullable` values for the same field across multiple indexes and in the format. The resulting value is the strictest one. For example, if even just one index or the format define `is_nullable = false`, the field is not nullable. It does not matter what is said in the other places of the space.

This, in turn, was done in order to support iterative DDL. It wasn't transactional in the beginning. And even now it is not stable when done in a transaction. Because of that if you wanted to change `is_nullable` you needed to drop all the indexes, change the format, and create the indexes back. Super expensive. Otherwise when you changed it in one index, the alter didn't work, because it conflicted with the other indexes.

With this feature you could change `is_nullable` in all the indexes and in the format step by step, in separate transactions. Without doing expensive index rebuilds.

The same comes to the field type. You can declare one field as 'unsigned', 'number', 'integer' and 'scalar' simultaneously, in different indexes and in the format. The resulting type would be 'unsigned' as the strictest one.

This is why it happens for this primary index thing. Just like it happens to any index. When you created the index, it took `is_nullable` from the format (`false`). Then you changed the format, but it didn't change the index - it remained `is_nullable = false` in the index definition. Hence the field remained non-nullable and all works fine, except that the space format might confuse you a bit.

Technically, primary index is not alterable. You can't change its parts. Which means, in theory, that we could ban nullable on PK parts specifically. But I see alter allows some SQL-related changes, which would conflict a bit with the idea of PK being non-alterable.

Final point is that if we disallow that usage, we might get backward compatibility issues. The instance having such a configuration won't bootstrap. Because during `_space` and `_index` recovery, it will notice the primary parts are alterable, but the new version won't allow that, and the instance won't start. It does not look like a win.

**Comment by alyapunov** [comment]
@Gerold103 thanks, almost the same as I said above except one thing: as I see a primary index is alterable.

**Comment by unera** [comment]
Current situation is the following:

Space format and indexes can be different.

For example, you can have format `{type=number}` and create three indexes: `unsigned`, `int`, `number`.
Each index will check tupple data before inserting, so you can't insert anything except `unsigned`.

I doubt that it is worth changing the (current) proceeding this case.
